runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractContextStreamView.java

```diff
@@ -135,6 +134,13 @@ public final synchronized ILogData nextUpTo(final long maxGlobal) {
             // Update the pointer.
             updatePointer(entry);
 
+            // We added hole to StreamView layer, in order to enable VLO sync to a hole
+            // and in this way, we can avoid unnecessary sync that call sequencer every time.
+            // It can expose a hole to sv consumer, so check if entry is a hole.
+            if (entry.isHole()) {
+                return nextUpTo(maxGlobal);
```

Comment:
Resursive call is not good practice in production, it could blow up stack. I know you are referencing from the recursive code right below, but that is basically dead code.

Also, why don't we do this in `removeFromQueue`?  We can read the next one if current is a hole in `removeFromQueue` right? That seems cleaner.

Reply:
Yes, I agree that recursive call is not good.

If we have a hole in streamA's tail, we read this hole can continue to read the next one in removeFromQueue, and return null. Then it won't call updatePointer(entry), so pointer is still behind this hole and we do not improve performance.

Reply:
This the problem of existing code organization, we need to clean it up later.
Though I think this does not matter in your case, as long as you put the hole into the resolvedQueue via `addToResolvedQueue`, the maxResolution will be updated: https://github.com/CorfuDB/CorfuDB/blob/master/runtime/src/main/java/org/corfudb/runtime/view/stream/AbstractQueuedStreamView.java#L87
 then next time maxResolution will prevent extra sequencer call. But using globalPointer instead of maxResolution can prevent unnecessary TrimmedException, so it's ok for now. I think later on we should reorg this part of code.